File: echo/server-starnet/Actor/GatewayService.cs
using System.Collections.Concurrent;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using ServerCs.Protocol;

namespace ServerCs.Actor;

public delegate string RouteHandler(Session session, JsonElement body);

public class GatewayService : Service
{
    private static readonly ConcurrentDictionary<string, RouteHandler> Handlers = new();
    private static readonly object HandlersLock = new();

    private readonly ConcurrentDictionary<int, Session> _sessions = new();
    private readonly object _sessionsLock = new();

    private volatile bool _heartbeatRunning = true;
    private Thread? _heartbeatThread;

    public GatewayService()
    {
        _heartbeatThread = new Thread(() =>
        {
            while (_heartbeatRunning)
            {
                Thread.Sleep(1000);
                CheckHeartbeatTimeout();
            }
        });
        _heartbeatThread.Start();
    }

    public static void RegisterHandler(string route, RouteHandler handler)
    {
        lock (HandlersLock)
        {
            Handlers[route] = handler;
        }
    }

    public override void OnInit()
    {
        Console.WriteLine($"[GatewayService] OnInit id={Id}");
    }

    public override void OnMsg(BaseMsg msg)
    {
        if (msg.Type == MsgType.SocketAccept)
        {
            if (msg is SocketAcceptMsg acceptMsg)
            {
                OnAcceptMsg(acceptMsg);
            }
        }
        else if (msg.Type == MsgType.SocketRW)
        {
            if (msg is SocketRWMsg rwMsg)
            {
                OnRWMsg(rwMsg);
            }
        }
    }

    public override void OnExit()
    {
        Console.WriteLine($"[GatewayService] OnExit id={Id}");
        _heartbeatRunning = false;
        _heartbeatThread?.Join();

        lock (_sessionsLock)
        {
            foreach (var pair in _sessions)
            {
                Starnet.Instance.CloseConn(pair.Key);
            }
            _sessions.Clear();
        }
    }

    protected override void OnAcceptMsg(SocketAcceptMsg msg)
    {
        int clientFd = msg.ClientFd;
        Console.WriteLine($"[GatewayService] OnAcceptMsg clientFd={clientFd}");

        var session = new Session
        {
            Fd = clientFd,
            State = ConnectionState.Inited,
            LastHeartbeat = DateTime.UtcNow
        };

        _sessions[clientFd] = session;
    }

    protected override void OnRWMsg(SocketRWMsg msg)
    {
        int fd = msg.Fd;
        if (msg.IsRead)
        {
            const int BUFFSIZE = 4096;
            byte[] buff = new byte[BUFFSIZE];
            int len = 0;

            try
            {
                var socket = Starnet.Instance.GetSocket(fd);
                if (socket == null) return;

                do
                {
                    len = socket.Receive(buff, 0, BUFFSIZE, SocketFlags.None);
                    if (len > 0)
                    {
                        OnSocketData(fd, buff, len);
                    }
                } while (len == BUFFSIZE && socket.Available > 0);
            }
            catch (SocketException ex)
            {
                if (ex.SocketErrorCode != SocketError.WouldBlock &&
                    ex.SocketErrorCode != SocketError.TimedOut)
                {
                    if (_sessions.ContainsKey(fd))
                    {
                        OnSocketClose(fd);
                        Starnet.Instance.CloseConn(fd);
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[GatewayService] Read error: {ex.Message}");
                if (_sessions.ContainsKey(fd))
                {
                    OnSocketClose(fd);
                    Starnet.Instance.CloseConn(fd);
                }
            }
        }

        if (msg.IsWrite)
        {
            if (_sessions.ContainsKey(fd))
            {
                OnSocketWritable(fd);
            }
        }
    }

    protected override void OnSocketData(int fd, byte[] buff, int len)
    {
        Session? session;
        lock (_sessionsLock)
        {
            if (!_sessions.TryGetValue(fd, out session))
            {
                Console.WriteLine($"[GatewayService] Session not found for fd={fd}");
                return;
            }
            session.DataBuf.AddRange(buff.Take(len));
        }

        // Process complete packages
        while (true)
        {
            List<byte>? pkgData = null;
            bool hasPackage = false;

            lock (_sessionsLock)
            {
                if (!_sessions.TryGetValue(fd, out session))
                {
                    return;
                }

                if (session.DataBuf.Count >= 4)
                {
                    int pkgType = session.DataBuf[0];
                    int pkgLen = (session.DataBuf[1] << 16) | (session.DataBuf[2] << 8) | session.DataBuf[3];
                    int totalLen = 4 + pkgLen;

                    if (session.DataBuf.Count >= totalLen)
                    {
                        pkgData = session.DataBuf.Take(totalLen).ToList();
                        session.DataBuf.RemoveRange(0, totalLen);
                        hasPackage = true;
                    }
                }
            }

            if (!hasPackage)
            {
                break;
            }

            if (pkgData != null)
            {
                var pkg = Package.Decode(pkgData.ToArray());
                if (pkg != null)
                {
                    ProcessPackage(fd, pkg);
                }
                else
                {
                    Console.WriteLine("[GatewayService] Failed to decode package");
                }
            }
        }
    }

    protected override void OnSocketWritable(int fd)
    {
        // Handle write buffer if needed
    }

    protected override void OnSocketClose(int fd)
    {
        Console.WriteLine($"[GatewayService] OnSocketClose fd={fd}");
        _sessions.TryRemove(fd, out _);
    }

    private void ProcessPackage(int fd, Package pkg)
    {
        switch (pkg.Type)
        {
            case PackageType.Handshake:
                HandleHandshake(fd, pkg.Body);
                break;
            case PackageType.HandshakeAck:
                HandleHandshakeAck(fd);
                break;
            case PackageType.Heartbeat:
                HandleHeartbeat(fd);
                break;
            case PackageType.Data:
                HandleData(fd, pkg.Body);
                break;
            case PackageType.Kick:
                OnSocketClose(fd);
                Starnet.Instance.CloseConn(fd);
                break;
            default:
                Console.WriteLine($"[GatewayService] Unknown package type: {pkg.Type}");
                break;
        }
    }

    private void HandleHandshake(int fd, byte[] body)
    {
        Console.WriteLine($"[GatewayService] handle_handshake fd={fd}");
        string response = "{\"code\":200,\"sys\":{\"heartbeat\":10,\"dict\":{},\"protos\":{\"client\":{},\"server\":{}}},\"user\":{}}";
        byte[] responseBody = Encoding.UTF8.GetBytes(response);
        byte[] responsePkg = Package.Encode(PackageType.Handshake, responseBody);
        Console.WriteLine($"[GatewayService] Sending handshake response, size={responsePkg.Length}");
        Send(fd, responsePkg);

        lock (_sessionsLock)
        {
            if (_sessions.TryGetValue(fd, out var session))
            {
                session.State = ConnectionState.WaitAck;
                session.HeartbeatInterval = TimeSpan.FromSeconds(10);
                session.HeartbeatTimeout = TimeSpan.FromSeconds(20);
                Console.WriteLine("[GatewayService] Handshake response sent, state=WaitAck");
            }
        }
    }

    private void HandleHandshakeAck(int fd)
    {
        lock (_sessionsLock)
        {
            if (_sessions.TryGetValue(fd, out var session))
            {
                session.State = ConnectionState.Working;
                session.LastHeartbeat = DateTime.UtcNow;
            }
        }
    }

    private void HandleHeartbeat(int fd)
    {
        lock (_sessionsLock)
        {
            if (_sessions.TryGetValue(fd, out var session))
            {
                session.LastHeartbeat = DateTime.UtcNow;
            }
        }

        byte[] heartbeatPkg = Package.Encode(PackageType.Heartbeat, null);
        Send(fd, heartbeatPkg);
    }

    private void HandleData(int fd, byte[] body)
    {
        lock (_sessionsLock)
        {
            if (_sessions.TryGetValue(fd, out var session))
            {
                session.LastHeartbeat = DateTime.UtcNow;
            }
        }

        var msg = Message.Decode(body);
        if (msg == null)
        {
            Console.WriteLine($"[GatewayService] Failed to decode message, body_size={body.Length}");
            return;
        }

        string msgBody = msg.Body.Length > 0 ? Encoding.UTF8.GetString(msg.Body) : "{}";
        if (msg.Type == MessageType.Request)
        {
            HandleRequest(fd, msg.Id, msg.Route, msgBody);
        }
        else if (msg.Type == MessageType.Notify)
        {
            Console.WriteLine($"[GatewayService] Notify received: route={msg.Route}, body={msgBody}");
        }
        else
        {
            Console.WriteLine($"[GatewayService] Unknown message type: {msg.Type}");
        }
    }

    private void HandleRequest(int fd, int id, string route, string body)
    {
        Console.WriteLine($"[GatewayService] handle_request fd={fd}, id={id}, route={route}, body={body}");
        string responseBody;

        lock (HandlersLock)
        {
            if (Handlers.TryGetValue(route, out var handler))
            {
                Session? session;
                lock (_sessionsLock)
                {
                    if (!_sessions.TryGetValue(fd, out session))
                    {
                        Console.WriteLine($"[GatewayService] Session not found for fd={fd}");
                        return;
                    }
                }

                JsonElement bodyJson;
                try
                {
                    if (!string.IsNullOrEmpty(body))
                    {
                        bodyJson = JsonDocument.Parse(body).RootElement;
                    }
                    else
                    {
                        bodyJson = JsonDocument.Parse("{}").RootElement;
                    }
                }
                catch (JsonException ex)
                {
                    Console.WriteLine($"[GatewayService] Failed to parse JSON body: {ex.Message}");
                    string errorResponseBody = "{\"code\":400,\"msg\":\"Invalid JSON\"}";
                    byte[] errorResponseBytes = Encoding.UTF8.GetBytes(errorResponseBody);
                    byte[] errorResponseMsg = Message.Encode(id, MessageType.Response, false, "", errorResponseBytes);
                    byte[] errorResponsePkg = Package.Encode(PackageType.Data, errorResponseMsg);
                    Console.WriteLine($"[GatewayService] Sending error response, pkg_size={errorResponsePkg.Length}");
                    Send(fd, errorResponsePkg);
                    return;
                }

                responseBody = handler(session, bodyJson);
            }
            else
            {
                Console.WriteLine($"[GatewayService] Unknown route: {route}");
                responseBody = $"{{\"code\":404,\"msg\":\"Route not found: {route}\"}}";
            }
        }

        byte[] responseBytes = Encoding.UTF8.GetBytes(responseBody);
        byte[] responseMsg = Message.Encode(id, MessageType.Response, false, "", responseBytes);
        byte[] responsePkg = Package.Encode(PackageType.Data, responseMsg);
        Send(fd, responsePkg);
    }

    private void Send(int fd, byte[] data)
    {
        try
        {
            var socket = Starnet.Instance.GetSocket(fd);
            if (socket == null) return;

            int sent = 0;
            while (sent < data.Length)
            {
                try
                {
                    int n = socket.Send(data, sent, data.Length - sent, SocketFlags.None);
                    if (n <= 0)
                    {
                        Console.WriteLine($"[GatewayService] Send returned 0, fd={fd}");
                        OnSocketClose(fd);
                        Starnet.Instance.CloseConn(fd);
                        return;
                    }
                    sent += n;
                }
                catch (SocketException ex)
                {
                    if (ex.SocketErrorCode == SocketError.WouldBlock)
                    {
                        Console.WriteLine($"[GatewayService] Send would block, fd={fd}, sent={sent}/{data.Length}");
                        // TODO: Add to write buffer and enable EPOLLOUT
                        return;
                    }
                    Console.WriteLine($"[GatewayService] Send error: {ex.SocketErrorCode}, fd={fd}");
                    OnSocketClose(fd);
                    Starnet.Instance.CloseConn(fd);
                    return;
                }
            }
        }
        catch (SocketException ex)
        {
            if (ex.SocketErrorCode == SocketError.WouldBlock)
            {
                Console.WriteLine($"[GatewayService] Send would block, fd={fd}");
                return;
            }
            Console.WriteLine($"[GatewayService] Send error: {ex.Message}, fd={fd}");
            OnSocketClose(fd);
            Starnet.Instance.CloseConn(fd);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[GatewayService] Send error: {ex.Message}, fd={fd}");
            OnSocketClose(fd);
            Starnet.Instance.CloseConn(fd);
        }
    }

    private void SendHeartbeat(int fd)
    {
        byte[] heartbeatPkg = Package.Encode(PackageType.Heartbeat, null);
        Send(fd, heartbeatPkg);
    }

    private void CheckHeartbeatTimeout()
    {
        var now = DateTime.UtcNow;
        var timeoutFds = new List<int>();

        lock (_sessionsLock)
        {
            foreach (var pair in _sessions)
            {
                var session = pair.Value;
                if (session.State == ConnectionState.Working)
                {
                    var elapsed = now - session.LastHeartbeat;
                    if (elapsed > session.HeartbeatTimeout)
                    {
                        timeoutFds.Add(pair.Key);
                    }
                    else if (elapsed >= session.HeartbeatInterval)
                    {
                        SendHeartbeat(pair.Key);
                    }
                }
            }
        }

        foreach (int fd in timeoutFds)
        {
            Console.WriteLine($"[GatewayService] Heartbeat timeout fd={fd}");
            OnSocketClose(fd);
            Starnet.Instance.CloseConn(fd);
        }
    }
}

File: echo/server-sunnet/src/GatewayService.cpp
#include "GatewayService.h"
#include "SessionManager.h"
#include "../sunnet/include/Sunnet.h"
#include <iostream>
#include <sys/socket.h>
#include <unistd.h>
#include <errno.h>
#include <cstring>
#include <chrono>
#include <thread>

using namespace std;

unordered_map<string, RouteHandler> GatewayService::handlers_;
mutex GatewayService::handlers_mutex_;

// 注册 GatewayService 到服务工厂（需要在 main 函数中调用）
void GatewayService::register_service() {
    Sunnet::RegisterService("gateway", []() -> shared_ptr<Service> {
        return make_shared<GatewayService>();
    });
}

void GatewayService::register_handler(const string& route, RouteHandler handler) {
    lock_guard<mutex> lock(handlers_mutex_);
    handlers_[route] = move(handler);
}

GatewayService::GatewayService() {
    // Start heartbeat thread
    heartbeat_thread_ = thread([this]() {
        while (heartbeat_running_) {
            this_thread::sleep_for(chrono::seconds(1));
            check_heartbeat_timeout();
        }
    });
}

GatewayService::~GatewayService() {
    heartbeat_running_ = false;
    if (heartbeat_thread_.joinable()) {
        heartbeat_thread_.join();
    }
}

void GatewayService::OnInit() {
    cout << "[GatewayService] OnInit id=" << id << endl;
}

void GatewayService::OnMsg(shared_ptr<BaseMsg> msg) {
    if (msg->type == BaseMsg::TYPE::SOCKET_ACCEPT) {
        auto m = dynamic_pointer_cast<SocketAcceptMsg>(msg);
        OnAcceptMsg(m);
    } else if (msg->type == BaseMsg::TYPE::SOCKET_RW) {
        auto m = dynamic_pointer_cast<SocketRWMsg>(msg);
        OnRWMsg(m);
    }
}

void GatewayService::OnExit() {
    cout << "[GatewayService] OnExit id=" << id << endl;
    lock_guard<mutex> lock(sessions_mutex_);
    for (auto& pair : sessions_) {
        Sunnet::inst->CloseConn(pair.first);
    }
    sessions_.clear();
}

void GatewayService::OnAcceptMsg(shared_ptr<SocketAcceptMsg> msg) {
    int clientFd = msg->clientFd;
    cout << "[GatewayService] OnAcceptMsg clientFd=" << clientFd << endl;

    auto session = make_shared<Session>();
    session->fd = clientFd;
    session->state = ConnectionState::Inited;
    session->last_heartbeat = chrono::steady_clock::now();

    lock_guard<mutex> lock(sessions_mutex_);
    sessions_[clientFd] = session;
}

void GatewayService::OnRWMsg(shared_ptr<SocketRWMsg> msg) {
    int fd = msg->fd;
    if (msg->isRead) {
        const int BUFFSIZE = 4096;
        char buff[BUFFSIZE];
        int len = 0;
        do {
            len = read(fd, &buff, BUFFSIZE);
            if (len > 0) {
                OnSocketData(fd, buff, len);
            }
        } while (len == BUFFSIZE);

        if (len <= 0 && errno != EAGAIN) {
            lock_guard<mutex> lock(sessions_mutex_);
            if (sessions_.find(fd) != sessions_.end()) {
                OnSocketClose(fd);
                Sunnet::inst->CloseConn(fd);
            }
        }
    }
    if (msg->isWrite) {
        lock_guard<mutex> lock(sessions_mutex_);
        if (sessions_.find(fd) != sessions_.end()) {
            OnSocketWritable(fd);
        }
    }
}

void GatewayService::OnSocketData(int fd, const char *buff, int len) {
    // Get session and copy data buffer
    shared_ptr<Session> session;
    {
        lock_guard<mutex> lock(sessions_mutex_);
        auto it = sessions_.find(fd);
        if (it == sessions_.end()) {
            cout << "[GatewayService] Session not found for fd=" << fd << endl;
            return;
        }
        session = it->second;
        session->data_buf.insert(session->data_buf.end(), buff, buff + len);
    }

    // Process complete packages (without holding lock)
    while (true) {
        vector<uint8_t> pkg_data;
        bool has_package = false;
        
        {
            lock_guard<mutex> lock(sessions_mutex_);
            auto it = sessions_.find(fd);
            if (it == sessions_.end()) {
                return;
            }
            session = it->second;
            
            if (session->data_buf.size() >= 4) {
                int pkg_type = session->data_buf[0];
                int pkg_len = (session->data_buf[1] << 16) | (session->data_buf[2] << 8) | session->data_buf[3];
                size_t total_len = 4 + pkg_len;

                if (session->data_buf.size() >= total_len) {
                    pkg_data.assign(session->data_buf.begin(), session->data_buf.begin() + total_len);
                    session->data_buf.erase(session->data_buf.begin(), session->data_buf.begin() + total_len);
                    has_package = true;
                }
            }
        }

        if (!has_package) {
            break;
        }

        auto pkg = protocol::Package::decode(pkg_data);
        if (pkg) {
            process_package(fd, *pkg);
        } else {
            cout << "[GatewayService] Failed to decode package" << endl;
        }
    }
}

void GatewayService::OnSocketWritable(int fd) {
    // Handle write buffer if needed
}

void GatewayService::OnSocketClose(int fd) {
    cout << "[GatewayService] OnSocketClose fd=" << fd << endl;
    lock_guard<mutex> lock(sessions_mutex_);
    bool had_session = (sessions_.erase(fd) > 0);
    
    // Notify session manager of connection close (only if session existed)
    if (had_session) {
        SessionManager::getInstance().onConnectionClose();
    }
}

void GatewayService::process_package(int fd, const protocol::Package& pkg) {
    switch (pkg.type) {
        case protocol::PackageType::Handshake:
            handle_handshake(fd, pkg.body);
            break;
        case protocol::PackageType::HandshakeAck:
            handle_handshake_ack(fd);
            break;
        case protocol::PackageType::Heartbeat:
            handle_heartbeat(fd);
            break;
        case protocol::PackageType::Data:
            handle_data(fd, pkg.body);
            break;
        case protocol::PackageType::Kick:
            OnSocketClose(fd);
            Sunnet::inst->CloseConn(fd);
            break;
        default:
            cout << "[GatewayService] Unknown package type: " << (int)static_cast<uint8_t>(pkg.type) << endl;
            break;
    }
}

void GatewayService::handle_handshake(int fd, const vector<uint8_t>& /*body*/) {
    cout << "[GatewayService] handle_handshake fd=" << fd << endl;
    string response = R"({"code":200,"sys":{"heartbeat":10,"dict":{},"protos":{"client":{},"server":{}}},"user":{}})";
    vector<uint8_t> response_body(response.begin(), response.end());
    auto response_pkg = protocol::Package::encode(protocol::PackageType::Handshake, response_body);
    cout << "[GatewayService] Sending handshake response, size=" << response_pkg.size() << endl;
    send(fd, response_pkg);

    lock_guard<mutex> lock(sessions_mutex_);
    auto it = sessions_.find(fd);
    if (it != sessions_.end()) {
        it->second->state = ConnectionState::WaitAck;
        it->second->heartbeat_interval = chrono::seconds(10);
        it->second->heartbeat_timeout = chrono::seconds(20);
        cout << "[GatewayService] Handshake response sent, state=WaitAck" << endl;
    }
}

void GatewayService::handle_handshake_ack(int fd) {
    lock_guard<mutex> lock(sessions_mutex_);
    auto it = sessions_.find(fd);
    if (it != sessions_.end()) {
        it->second->state = ConnectionState::Working;
        it->second->last_heartbeat = chrono::steady_clock::now();
        
        // Notify session manager of successful handshake
        SessionManager::getInstance().onHandshakeSuccess();
    }
}

void GatewayService::handle_heartbeat(int fd) {
    lock_guard<mutex> lock(sessions_mutex_);
    auto it = sessions_.find(fd);
    if (it != sessions_.end()) {
        it->second->last_heartbeat = chrono::steady_clock::now();
    }

    auto heartbeat_pkg = protocol::Package::encode(protocol::PackageType::Heartbeat, {});
    send(fd, heartbeat_pkg);
}

void GatewayService::handle_data(int fd, const vector<uint8_t>& body) {
    {
        lock_guard<mutex> lock(sessions_mutex_);
        auto it = sessions_.find(fd);
        if (it != sessions_.end()) {
            it->second->last_heartbeat = chrono::steady_clock::now();
        }
    }

    auto msg = protocol::Message::decode(body);
    if (!msg) {
        cout << "[GatewayService] Failed to decode message, body_size=" << body.size() << endl;
        return;
    }

    string msg_body(msg->body.begin(), msg->body.end());
    if (msg->type == protocol::MessageType::Request) {
        handle_request(fd, msg->id, msg->route, msg_body);
    } else if (msg->type == protocol::MessageType::Notify) {
        cout << "[GatewayService] Notify received: route=" << msg->route << ", body=" << msg_body << endl;
    } else {
        cout << "[GatewayService] Unknown message type: " << (int)static_cast<int>(msg->type) << endl;
    }
}

void GatewayService::handle_request(int fd, int id, const string& route, const string& body) {
    cout << "[GatewayService] handle_request fd=" << fd << ", id=" << id << ", route=" << route << ", body=" << body << endl;
    string response_body;

    lock_guard<mutex> lock(handlers_mutex_);
    auto it = handlers_.find(route);
    if (it != handlers_.end()) {
        lock_guard<mutex> session_lock(sessions_mutex_);
        auto session_it = sessions_.find(fd);
        if (session_it == sessions_.end()) {
            cout << "[GatewayService] Session not found for fd=" << fd << endl;
            return;
        }
        auto session = session_it->second;

        json body_json;
        try {
            if (!body.empty()) {
                body_json = json::parse(body);
            }
        } catch (const json::parse_error& e) {
            cout << "[GatewayService] Failed to parse JSON body: " << e.what() << endl;
            response_body = R"({"code":400,"msg":"Invalid JSON"})";
            vector<uint8_t> response_bytes(response_body.begin(), response_body.end());
            auto response_msg = protocol::Message::encode(id, protocol::MessageType::Response, false, "", response_bytes);
            auto response_pkg = protocol::Package::encode(protocol::PackageType::Data, response_msg);
            cout << "[GatewayService] Sending error response, pkg_size=" << response_pkg.size() << endl;
            send(fd, response_pkg);
            return;
        }

        response_body = it->second(*session, body_json);
    } else {
        cout << "[GatewayService] Unknown route: " << route << endl;
        response_body = R"({"code":404,"msg":"Route not found: )" + route + R"("})";
    }

    vector<uint8_t> response_bytes(response_body.begin(), response_body.end());
    auto response_msg = protocol::Message::encode(id, protocol::MessageType::Response, false, "", response_bytes);
    auto response_pkg = protocol::Package::encode(protocol::PackageType::Data, response_msg);
    send(fd, response_pkg);
}

void GatewayService::send(int fd, const vector<uint8_t>& data) {
    ssize_t sent = 0;
    while (sent < static_cast<ssize_t>(data.size())) {
        ssize_t n = ::send(fd, data.data() + sent, data.size() - sent, 0);
        if (n < 0) {
            if (errno == EAGAIN || errno == EWOULDBLOCK) {
                // Socket buffer is full, will retry later
                cout << "[GatewayService] Send would block, fd=" << fd << ", sent=" << sent << "/" << data.size() << endl;
                // TODO: Add to write buffer and enable EPOLLOUT
                return;
            }
            cout << "[GatewayService] Send error: " << strerror(errno) << ", fd=" << fd << endl;
            OnSocketClose(fd);
            Sunnet::inst->CloseConn(fd);
            return;
        }
        sent += n;
    }
}

void GatewayService::send_heartbeat(int fd) {
    auto heartbeat_pkg = protocol::Package::encode(protocol::PackageType::Heartbeat, {});
    send(fd, heartbeat_pkg);
}

void GatewayService::check_heartbeat_timeout() {
    auto now = chrono::steady_clock::now();
    vector<int> timeout_fds;

    {
        lock_guard<mutex> lock(sessions_mutex_);
        for (auto& pair : sessions_) {
            auto session = pair.second;
            if (session->state == ConnectionState::Working) {
                auto elapsed = chrono::duration_cast<chrono::seconds>(now - session->last_heartbeat);
                if (elapsed > session->heartbeat_timeout) {
                    timeout_fds.push_back(pair.first);
                } else if (elapsed >= session->heartbeat_interval) {
                    send_heartbeat(pair.first);
                }
            }
        }
    }

    for (int fd : timeout_fds) {
        cout << "[GatewayService] Heartbeat timeout fd=" << fd << endl;
        OnSocketClose(fd);
        Sunnet::inst->CloseConn(fd);
    }
}
